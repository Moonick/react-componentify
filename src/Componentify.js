// @flow
import React, { Component } from "react";

type Match = RegExp$matchResult;
type PropsObj = { [key: string]: { [key: string]: string } };

type PropsFunc = Match => { [key: string]: string };

type Converter = {
  regex: RegExp,
  component: string | (() => void),
  props?: PropsObj | PropsFunc,
  innerText?: Match => void,
  match?: Match
};

type ComponentifyProps = {
  text: string,
  converters: Array<Converter>
};

class Componentify extends Component<ComponentifyProps, {}> {
  static defaultProps = {
    text: "",
    converters: []
  };

  getPlainTextComponent(text: string, key?: string) {
    return <span key={key}>{text}</span>;
  }

  getCurrentConverter(text: string) {
    const { converters } = this.props;

    return converters.reduce((currtentConverter, converter) => {
      // Clone so mutating doesn't affect client-passed objects
      converter = Object.assign({}, converter);
      const regex = converter.regex;

      if (!regex) {
        throw new Error("Invalid regex");
      }

      const currentMatch = regex.exec(text);

      if (currentMatch !== null) {
        converter.match = currentMatch;
        const lowestIndex = currtentConverter && currtentConverter.match.index;
        const currentIndex = converter.index;

        if (currtentConverter === null || currentIndex < lowestIndex) {
          currtentConverter = converter;
        }
      }

      return currtentConverter;
    }, null);
  }

  generateComponent(converter: Converter, key: string): React$CreateElement {
    const { component, match } = converter;
    let { props } = converter;
    let { innerText } = converter;
    let children = null;

    if (typeof props === "function") {
      props = props(match);
    }

    if (props) {
      props.key = key;
    }

    if (typeof innerText === "function") {
      innerText = innerText(match);
    }

    if (typeof innerText !== "undefined") {
      children = this.generateComponentList(innerText, match[0]);
    }

    return React.createElement(component, props, children);
  }

  generateComponentList(text: string, prevMatch: string) {
    let str = text;
    let components = [];

    while (str !== "") {
      const currtentConverter = this.getCurrentConverter(str);

      if (!currtentConverter || prevMatch === currtentConverter.match[0]) {
        break;
      }

      const matchIndex = currtentConverter.match.index;
      const textBeforeMatch = str.slice(0, matchIndex);
      const textAfterMatch = str.slice(
        matchIndex + currtentConverter.match[0].length
      );
      str = textAfterMatch;

      if (textBeforeMatch !== "") {
        components.push(
          this.getPlainTextComponent(
            textBeforeMatch,
            components.length.toString()
          )
        );
      }

      components.push(
        this.generateComponent(currtentConverter, components.length.toString())
      );
    }

    if (str !== "") {
      if (prevMatch !== "") {
        components.push(str);
      } else {
        components.push(
          this.getPlainTextComponent(str, components.length.toString())
        );
      }
    }

    return components;
  }

  render() {
    const { text, converters } = this.props;

    if (!text) {
      throw new Error('Missing property "text"');
    }

    if (!converters) {
      return this.getPlainTextComponent(text);
    }

    return this.generateComponentList(text, "");
  }
}

export default Componentify;
