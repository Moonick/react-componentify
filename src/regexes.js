export const LINK_REGEX = /(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?(?:\[(.+)\])?/;
export const BOLD_REGEX = /\*([\w\d\s\:\/\.\[\]]+)\*/;
export const ITALIC_REGEX = /\_([\w\d\s\:\/\.\[\]]+)\_/;
export const BR_REGEX = /<br\/>/;
