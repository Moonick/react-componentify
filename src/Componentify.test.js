import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Componentify from "./Componentify";
import { ITALIC_REGEX } from "./regexes";
import {
  boldConverter,
  italicConverter,
  linkConverter,
  brTagConverter
} from "./converters";

const MyComponent = ({ children }) => {
  return <span style={{ fontStyle: "italic" }}>{children}</span>;
};

describe("<Componentify />", () => {
  describe("without nesting", () => {
    it("renders plain text", () => {
      const div = document.createElement("div");

      ReactDOM.render(<Componentify text={"plain text"} />, div);
      expect(div.innerHTML).toContain("plain text");
      ReactDOM.unmountComponentAtNode(div);
    });

    it("renders text with bold words", () => {
      const tree = renderer
        .create(
          <Componentify
            text={"Тhis is bold *text* and *more* text"}
            converters={[boldConverter]}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders text with italic words", () => {
      const tree = renderer
        .create(
          <Componentify
            text={"Тhis is italic _text_ more text"}
            converters={[italicConverter]}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    describe("text with link", () => {
      it("renders text with link with name", () => {
        const tree = renderer
          .create(
            <Componentify
              text={"Тhis is https://google.com[google] link"}
              converters={[linkConverter]}
            />
          )
          .toJSON();
        expect(tree).toMatchSnapshot();
      });

      it("renders text with link without name", () => {
        const tree = renderer
          .create(
            <Componentify
              text={"Тhis is https://google.com link"}
              converters={[linkConverter]}
            />
          )
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    it("renders text with <br /> tag", () => {
      const tree = renderer
        .create(
          <Componentify
            text={"Тhis is line one<br/>This is line two"}
            converters={[brTagConverter]}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders text with replaced words", () => {
      const tree = renderer
        .create(
          <Componentify
            text={"My name is John"}
            converters={[
              {
                regex: /John/,
                component: "span",
                innerText: "Snow"
              }
            ]}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("with nesting", () => {
    it("renders text with bold and italic words", () => {
      const tree = renderer
        .create(
          <Componentify
            text={"Тhis is my *bold and _italic_* text"}
            converters={[boldConverter, italicConverter]}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders text with bold link", () => {
      const tree = renderer
        .create(
          <Componentify
            text={"Тhis is my bold *https://google.com* link"}
            converters={[boldConverter, linkConverter]}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("with custom component", () => {
    it("renders text with italic", () => {
      const tree = renderer
        .create(
          <Componentify
            text={"Тhis is my italic _text_"}
            converters={[
              {
                regex: ITALIC_REGEX,
                component: MyComponent,
                props: ([_, text]) => {
                  return {
                    text: text
                  };
                },
                innerText: matches => matches[2] || matches[1]
              }
            ]}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
