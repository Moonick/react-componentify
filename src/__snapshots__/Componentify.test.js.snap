// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Componentify /> with custom component renders text with italic 1`] = `
Array [
  <span>
    Тhis is my italic 
</span>,
  <span
    style={
        Object {
            "fontStyle": "italic",
          }
    }
>
    text
</span>,
]
`;

exports[`<Componentify /> with nesting renders text with bold and italic words 1`] = `
Array [
  <span>
    Тhis is my 
</span>,
  <span
    style={
        Object {
            "fontWeight": "900",
          }
    }
>
    <span>
        bold and 
    </span>
    <span
        style={
            Object {
                "fontStyle": "italic",
              }
        }
    >
        italic
    </span>
</span>,
  <span>
     text
</span>,
]
`;

exports[`<Componentify /> with nesting renders text with bold link 1`] = `
Array [
  <span>
    Тhis is my bold 
</span>,
  <span
    style={
        Object {
            "fontWeight": "900",
          }
    }
>
    <a
        href="https://"
        targer="_blank"
    >
        google
    </a>
</span>,
  <span>
     link
</span>,
]
`;

exports[`<Componentify /> without nesting renders text with <br /> tag 1`] = `
Array [
  <span>
    Тhis is line one
</span>,
  <br />,
  <span>
    This is line two
</span>,
]
`;

exports[`<Componentify /> without nesting renders text with bold words 1`] = `
Array [
  <span>
    Тhis is bold 
</span>,
  <span
    style={
        Object {
            "fontWeight": "900",
          }
    }
>
    text
</span>,
  <span>
     and 
</span>,
  <span
    style={
        Object {
            "fontWeight": "900",
          }
    }
>
    more
</span>,
  <span>
     text
</span>,
]
`;

exports[`<Componentify /> without nesting renders text with italic words 1`] = `
Array [
  <span>
    Тhis is italic 
</span>,
  <span
    style={
        Object {
            "fontStyle": "italic",
          }
    }
>
    text
</span>,
  <span>
     more text
</span>,
]
`;

exports[`<Componentify /> without nesting renders text with replaced words 1`] = `
Array [
  <span>
    My name is 
</span>,
  <span>
    Snow
</span>,
]
`;

exports[`<Componentify /> without nesting text with link renders text with link with name 1`] = `
Array [
  <span>
    Тhis is 
</span>,
  <a
    href="https://"
    targer="_blank"
>
    google
</a>,
  <span>
     link
</span>,
]
`;

exports[`<Componentify /> without nesting text with link renders text with link without name 1`] = `
Array [
  <span>
    Тhis is 
</span>,
  <a
    href="https://"
    targer="_blank"
>
    google
</a>,
]
`;
